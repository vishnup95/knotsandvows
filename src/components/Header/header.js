import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import { imagePath } from '../../utils/assetUtils';
import * as actions from '../../reducers/session/actions';
import * as homeActions from '../../modules/home/actions'
import * as modalActions from '../../reducers/modal/actions';
import { Link } from 'react-router-dom';
import { push, replace } from 'connected-react-router';
import { sendGAEvent } from '../../utils/GAUtilities'
import queryString from 'query-string';
import TalkToWeddingPlanner from '../../components/TalkToWeddingPlanner/talkToWeddingPlanner';

import {
    Collapse,
    Navbar,
    NavbarToggler,
    NavbarBrand,
    Nav,
    NavItem,
    NavLink,
    Modal,
    UncontrolledDropdown,
    DropdownToggle,
    DropdownMenu,
    DropdownItem
} from 'reactstrap';
import styles from './header.scss';
import modalStyles from '../../modals/forgotPasswordModal/forgotPasswordModal.scss';
import SignInModal from '../../modals/signInModal/SignInModal';
import ForgotPassword from "../../modals/forgotPasswordModal/ForgotPasswordModal"
import { isLoggedIn } from '../../utils/utilities';

const mapStateToProps = state => ({
    route: state.router.location.pathname,
    user: state.session.user,
    apiStatus: state.session.apiStatus,
    showLogin: state.session.showLogin,
    showForgotPassword: state.session.showForgotPassword,
    showResetPassword: state.session.showResetPassword,
    categories: state.home.categories,
    location: state.router.location,
});

const mapDispatchToProps = dispatch => ({
    actions: bindActionCreators({ ...actions }),
    dispatch
});
class Header extends Component {
    constructor(props) {
        super(props);
        this.toggleModal = this.toggleModal.bind(this);
        this.toggle = this.toggle.bind(this);
        this.state = {
            isOpen: false,
            hashValue: null,
            email: null
        };

    }

    static fetchData(store) {
        // Normally you'd pass action creators to "connect" from react-redux,
        // but since this is a static method you don't have access to "this.props".

        // Dispatching actions from "static fetchData()" will look like this (make sure to return a Promise):
        return store.dispatch(homeActions.fetchCategories());
    }

    toggleModal() {
        if (this.props.showLogin) {
            this.props.dispatch(actions.hideLogin());
        } else {
            sendGAEvent("Header", "Show Login");
            this.props.dispatch(actions.showLogin());
        }
    }

    componentWillMount() {
        if (this.props.categories.length === 0) {
            this.props.dispatch(homeActions.fetchCategories());
        }
    }

    componentDidMount() {

        if (isLoggedIn()) {
            this.props.dispatch(actions.fetchMyProfile());
        } 
    }

    toggleForgotPasswordModal = () => {
        if (this.props.showForgotPassword) {
            this.props.dispatch(actions.hideForgotPassword());
        } else {
            this.props.dispatch(actions.showForgotPassword());
        }
    }

    toggleResetPasswordModal = () => {
        if (this.props.showResetPassword) {
            this.props.dispatch(actions.hideResetPassword());
            this.props.dispatch(replace("/"));
        } else {
            this.props.dispatch(actions.showResetPassword());
        }
    }

    toggle() {
        this.setState({
            isOpen: !this.state.isOpen
        });
    }

    renderCategoryLists(list) {
        const listItems = list.map((item, index) => {
            return (
                <li key={index}>
                    <Link to={`/categories/${item.page_name}`}>{item.name}</Link>
                </li>
            );
        });

        return listItems;
    }

    logout = () => {
        if(localStorage){
            localStorage.clear();
        }
        this.props.dispatch(actions.clearUserData());
        this.navigateTo("/");
    }

    shortName = (userName) => {
        let name = userName;
        var initials = name.match(/\b\w/g) || [];
        initials = ((initials.shift() || '') + (initials.pop() || '')).toUpperCase();
        return (initials);
    }

    componentDidUpdate(prevProps) {
        if (prevProps == undefined) {
            return false;
        }

        if (this.props.location.search != prevProps.location.search) {
            if (this.props.location.pathname === "/resetpassword") {
                var hashValue = queryString.parse(this.props.location.search).code;
                var email = queryString.parse(this.props.location.search).email;
                if (hashValue) {
                    this.setState({ hashValue: hashValue, email });
                    this.toggleResetPasswordModal();
                }
            }
            else if (this.props.location.pathname === "/verify") {
                const activationCode = queryString.parse(this.props.location.search).activation_code;
                const email = queryString.parse(this.props.location.search).email;
                if (activationCode && email) {
                    this.props.dispatch(actions.verifyEmail(activationCode, email));
                }
            }
            else if (queryString.parse(this.props.location.search).login){
                // const { from } = this.props.location.state || { from: { pathname: '/' } }
                var login = queryString.parse(this.props.location.search).login;
                if (login == "true" && !isLoggedIn()) {
                    this.toggleModal();
                }else{
                    this.props.dispatch(replace("/"));
                }
            }   
        }   
        
        if (this.props.location.pathname === "/verify") {
            if (this.props.apiStatus == true) {
                this.props.dispatch(modalActions.showModal(`Email Verified Successfully`));
                this.props.dispatch(replace("/"));
            } else if (this.props.apiStatus == false) {
                this.props.dispatch(replace("/"));
                this.props.dispatch(modalActions.showModal(`Link expired`));
            }
        }
    }

    renderLoginItem = () => {

        if (this.props.user !== null) {
            return (
                <div>
                    <UncontrolledDropdown nav inNavbar>
                        <DropdownToggle nav className={styles.iconLink} style={{ cursor: "pointer", alignItems: "flex-end" }}>
                            <span className={styles.userInfo}>
                                {this.shortName(this.props.user.name)}
                            </span>
                        </DropdownToggle>

                        <DropdownMenu className={styles.userDropdown} >
                            <DropdownItem className="text-center" onClick={() => this.navigateTo("/profile")}>
                                Profile
                        </DropdownItem>
                            <DropdownItem className="text-center">
                                My bookings
                        </DropdownItem>
                            <DropdownItem className="text-center" onClick={() => this.logout()}>
                                Logout
                        </DropdownItem>
                        </DropdownMenu>
                    </UncontrolledDropdown>
                </div>
            );
        } else {
            return (
                <NavItem>
                    <NavLink className={styles.iconLink} style={{ cursor: "pointer" }} onClick={this.toggleModal}>
                        <img src={imagePath('avatar.svg')} alt="avatar" />
                        Login / Sign Up
                </NavLink>
                </NavItem>
            );
        }
    }

    navigateTo(route) {
        this.props.dispatch(push(route));

    }

    render() {

        return (
            <div className={`${styles.ahHeader}`}>
                <div className={styles.navSmall}>

                    <NavbarBrand href="/">
                        <img className={styles.logoTest} src={imagePath('logo.svg')} alt="logo" />
                    </NavbarBrand>
                    <Nav className={`${styles.iconNav}`} navbar>
                        {/* <NavItem>
                            <NavLink href="" className={styles.iconLink}>
                                <img src={imagePath('vendor.svg')} alt="vendor" />
                                For Vendors
                                </NavLink>
                        </NavItem> */}
                        {this.renderLoginItem()}
                    </Nav>
                </div>
                <Navbar color="" expand="md" className={styles.ahNav}>
                    <NavbarToggler className={this.state.isOpen ? 'close-nav' : ''} onClick={this.toggle} />
                    <Collapse navbar className={`${styles.ahCollapse} ${this.state.isOpen ? 'show' : ''}`} >
                        <Nav className="" navbar>
                            <NavItem className={styles.vendors}>
                                <NavLink onClick={() => this.navigateTo('/categories')}>Vendors</NavLink>
                                <div className={styles.categoriesList}>
                                    <ul>{this.renderCategoryLists(this.props.categories).splice(0, 6)}</ul>
                                    {this.renderCategoryLists(this.props.categories).length > 6 &&

                                        <ul>{this.renderCategoryLists(this.props.categories).splice(6, 6)}</ul>
                                    }
                                    {this.renderCategoryLists(this.props.categories).length > 12 &&

                                        <ul>{this.renderCategoryLists(this.props.categories).splice(12, 6)}</ul>
                                    }
                                </div>
                            </NavItem>
                            <NavItem>
                                <NavLink onClick={() => this.navigateTo('/packages')}>Packages</NavLink>
                            </NavItem>

                            <NavItem>
                                <NavLink onClick={() => this.navigateTo('/wishlist')}>Wishlist</NavLink>
                            </NavItem>

                            <NavItem>
                                <NavLink onClick={() => this.navigateTo('/services')}>Services</NavLink>
                            </NavItem>
                            <NavItem>
                                <NavLink onClick={() => this.navigateTo('/about')}>About</NavLink>
                            </NavItem>

                        </Nav>
                    </Collapse>
                </Navbar>
                <Modal isOpen={this.props.showLogin} toggle={this.toggleModal} centered={true} className={styles.loginModal}>
                    <SignInModal close={this.toggleModal} showForgotPassword={this.toggleForgotPasswordModal}></SignInModal>
                </Modal>
                <Modal isOpen={this.props.showForgotPassword} className={modalStyles.forgotContainer} toggle={this.toggleForgotPasswordModal} centered={true}>
                    <ForgotPassword></ForgotPassword>
                </Modal>
                <Modal isOpen={this.props.showResetPassword} toggle={this.toggleResetPasswordModal} className={modalStyles.forgotContainer} centered={true}>
                    <ForgotPassword hash={this.state.hashValue} email={this.state.email}></ForgotPassword>
                </Modal>
                <div className={styles.talkBtn}>
                    <TalkToWeddingPlanner type={'call'} />
                </div>
            </div>
        );
    }

}

Header.propTypes = {
    route: PropTypes.string,
    user: PropTypes.object,
    dispatch: PropTypes.func,
    actions: PropTypes.object,
    showLogin: PropTypes.bool,
    showForgotPassword: PropTypes.bool,
    showResetPassword: PropTypes.bool,
    history: PropTypes.any,
    categories: PropTypes.array,
    location: PropTypes.object,
    apiStatus: PropTypes.object
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Header);
