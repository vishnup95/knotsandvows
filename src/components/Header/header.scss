@import '../../variables/fonts';
@import '../../variables/colors';
@import '../../variables/utils';

.logo-test {
  max-height: 4.8rem;
  @include tablet {
    max-height: 8.1rem;
  }
}

.ah-header {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  z-index: 100;
  font-family: $roboto-regular;
  transition: 300ms all ease;
  background: $white;

  .ahNav {
    background-color: transparent !important;
    margin-bottom: 0;
    justify-content: space-between !important;
    height: 73px;
    padding: 0 5%;
    @include tablet {
      height: 8.1rem;
    }
    @include desktop {
      padding: 0 10%;
    }

    ul {
      @include tablet {
         @include align-center();
      }
    }
  }
}

.nav-small {
  @include mobile {
    display: none;
  }
  height: 48px;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 0 5rem 0 1rem;
  background: $medium-pink;
  @include tablet {
    padding: 0 8.333333%;
  }

}

.ahCollapse {
  @include mobile {
    position: fixed;
    top: 83px;
    left: 0;
  }
  width: auto;
  background-color: $white;
  box-shadow: 0 3px 6px 0 $black-16;
  padding: 1.6875rem 5.25rem 1.6875rem 2.25rem;
  border-radius: 0 5px 5px 0;
  flex-grow: 0 !important;
  @include tablet {
    height: 100%;
    position: relative;
    box-shadow: none;
    padding: 0;
    justify-content: center;
    border-radius: 0;
  }

  > ul {
    height: 100%;

    > li {
      height: 100%;
      @include flex();
      @include align-center();
      position: relative;

      a {
        cursor: pointer;
        @include mobile {
          line-height: 2.36;
          font-size: 18px;
          color: $greyish-brown-six !important;
          font-family: $roboto-regular;

        }
        font-family: $roboto-medium;
        color: $greyish-brown-six !important;
        opacity: .95;
        position: relative;
        @include fontSize(20.5);
        @include tablet {
          margin: 0 1rem;
          &:before {
            content: '';
            position: absolute;
            left: 20%;
            bottom: -1rem;
            width: 60%;
            background-color: $secondary-color;
            height: 0;
            border-radius: 2px;
          }

          &:hover {
            &:before {

              bottom: .4rem;
              height: 2px;
              transition: 300ms all ease;

            }
          }
        }
        @include desktop {
          margin: 0 2rem;
          }
      }

      a[class*=" active"] {
        color: $black;
      }

    }
  }

}

.icon-nav {
  height: 100%;
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  align-items: center;
  @include mobile {
    display: none;
  }

  > li {
    height: 100%;
    display: flex;
    align-items: center;

    &:last-child {
      .icon-link {
        @include mobile {
          border-right: 1px solid $black-16;
        }
      }
    }
  }

  .icon-link {
    @include flex();
    @include flex-column();
    @include justify-center();
    font-size: .875rem !important;
    color: $black-six !important;
    background-color: rgba(255,255,255,.5);
    height: calc(100% - 1rem);
    width: 113px / 16px * 1rem;
    @include align-center();
    margin: 0 1px !important;

    &:before {
      display: none;
    }

    > img {
      margin-bottom: .4rem;
      @include height(30);
    }
  }
}

.icon-link {
  padding: .5rem 0 !important;
}

.categoriesList {
  position: absolute;
  top: 15rem;
  left: 0;
  padding: 0;
  list-style: none;
  box-shadow: 0 1px 3px rgba(0,0,0,.12), 0 1px 2px rgba(0,0,0,.12);
  background-color: $white;
  height: 0;
  overflow: hidden;
  opacity: 0;
  display: flex;
  @include borderRadius(0,0,18,18);

  ul {
    padding: 0 !important;

    li {
      @extend %pointer;
      padding: 1rem 2.5rem 1.5rem;
      list-style: none;
      white-space: nowrap;

      a {
        color: $black-fourteen !important;
        font-family: $roboto-regular;

        &:before {
          content: '';
          position: absolute;
          left: 15%;
          bottom: -1rem;
          width: 70%;
          background-color: $secondary-color;
          height: 0;
          border-radius: 2px;
        }

        &:hover {
          text-decoration: none;

          &:before {
            bottom: -.4rem;
            height: 2px;
            transition: 300ms all ease;

          }
        }
      }
    }
  }
  @include mobile {
    display: none;
  }
}

.vendors {
  &:hover {
    .categoriesList {
      height: auto;
      top: 50px;
      transition: 500ms all ease;
      opacity: 1;
      @include padding(0,0,16);
    }
  }
}

.login-modal {
  @include width(1385);
}

.user-info {
  background-color: $primary-color;
  @include width(48);
  @include height(48);
  border-radius: 50%;
  display: flex;
  align-items: center;
  justify-content: center;
  font-family: $roboto-bold;
  @include fontSize(18);
  color: $white;
  letter-spacing: .5px;
}

.userDropdown {
  position: absolute !important;
  left: auto;
  right: -1.25rem;
  top: 4.7rem;
  box-shadow: 0 2px 6px 0 rgba(0, 0, 0, 0.3);
  border: none;
  @include borderRadius(8);

  &:before {
    content: '';
    height: .7rem;
    width: .7rem;
    background: $white;
    position: absolute;
    top: -0.25rem;
    right: 25%;
    transform: rotate(45deg);
  }

  button {
    cursor: pointer;
    padding: .7rem;
    font-family: $roboto-light;

    &:active,
    &:focus,
    &:hover {
      background-color: #ee4d9b29;
      outline: 0;
    }
  }
}

.talk-btn {
  position: fixed;
  bottom: 2rem;
  right: 2rem;
}

.mobileMenuModal {
  top: 5rem;
  left: 0;
  position: absolute;
  width: max-content;

  ul {
    padding: 20px;
    list-style: none;

    li {
      padding: 10px 6px;
    }
  }
}

.closeBtnSmall {
  position: absolute;
  top: 0;
  right: 5px;
  cursor: pointer;
  display: none;

  @include mobile {
    display: block;
  }

  img {
    @include width(28.3);
  }
}