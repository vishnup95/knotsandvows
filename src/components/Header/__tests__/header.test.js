import React from 'react';
import Header from '../header.js';
import { shallow } from 'enzyme';

describe('Header', () => {
  it('header renders without crashing', () => {
    const wrapper = shallow(
      <Header route="/"/>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});



