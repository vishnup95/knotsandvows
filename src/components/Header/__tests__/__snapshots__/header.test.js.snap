// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseButton header renders without crashing 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
